airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,170 @@
+    this.cdcStateHandler = cdcStateHandler;
+    this.targetPosition = targetPosition;
+    this.eventConverter = eventConverter;
+    this.offsetManager = offsetManager;
+    this.trackSchemaHistory = trackSchemaHistory;
+    this.schemaHistoryManager = schemaHistoryManager;
+    this.previousCheckpointOffset = (HashMap<String, String>) offsetManager.read();
+    this.initialOffset = new HashMap<>(this.previousCheckpointOffset);
+    resetCheckpointValues();
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(ConfiguredAirbyteStream stream) {
+    LOGGER.info("Sending CDC checkpoint state message.");
+    final AirbyteStateMessage stateMessage = createStateMessage(checkpointOffsetToSend);
+    previousCheckpointOffset.clear();
+    previousCheckpointOffset.putAll(checkpointOffsetToSend);
+    sendCheckpointMessage = false;
+    return stateMessage;
+  }
+
+  /**
+   * @param stream
+   * @param message
+   * @return
+   */
+  @Override
+  public AirbyteMessage processRecordMessage(ConfiguredAirbyteStream stream, ChangeEventWithMetadata message) {
+
+    if (checkpointOffsetToSend.isEmpty()) {
+      try {
+        final HashMap<String, String> temporalOffset = (HashMap<String, String>) offsetManager.read();
+        if (!targetPosition.isSameOffset(previousCheckpointOffset, temporalOffset)) {
+          checkpointOffsetToSend.putAll(temporalOffset);
+        }
+      } catch (final ConnectException e) {
+        LOGGER.warn("Offset file is being written by Debezium. Skipping CDC checkpoint in this loop.");
+      }
+    }
+
+    if (checkpointOffsetToSend.size() == 1
+        && !message.isSnapshotEvent()) {
+      if (targetPosition.isEventAheadOffset(checkpointOffsetToSend, message)) {
+        sendCheckpointMessage = true;
+      } else {
+        LOGGER.info("Encountered records with the same event offset.");
+      }
+    }
+
+    return eventConverter.toAirbyteMessage(message);
+  }
+
+  /**
+   * @param stream
+   * @return
+   */
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(ConfiguredAirbyteStream stream) {
+    final var syncFinishedOffset = (HashMap<String, String>) offsetManager.read();
+
+    System.out.println("initialoffset: " + initialOffset);
+    System.out.println("syncFinishedOffset: " + syncFinishedOffset);
+
+    if (targetPosition.isSameOffset(initialOffset, syncFinishedOffset)) {
+      // Edge case where no progress has been made: wrap up the
+      // sync by returning the initial offset instead of the
+      // current offset. We do this because we found that
+      // for some databases, heartbeats will cause Debezium to
+      // overwrite the offset file with a state which doesn't
+      // include all necessary data such as snapshot completion.
+      // This is the case for MS SQL Server, at least.
+      return createStateMessage(initialOffset);
+    }
```

Comment:
We might be able to get rid of this check entirely at this point. Try it out and see if source-mssql tests pass

cc @akashkulk

Reply:
it will break some test - it's minor though, turns out state in global state will have some minor changes:

before:
```
    "mssql_cdc_offset": {
      "[\"db_ziwfjgytty\",{\"server\":\"db_ziwfjgytty\",\"database\":\"db_ziwfjgytty\"}]": "{\"commit_lsn\":\"00000029:00000298:0003\",\"snapshot\":true,\"snapshot_completed\":true}"
    },
```

after:
```
    "mssql_cdc_offset": {
      "[\"db_ziwfjgytty\",{\"server\":\"db_ziwfjgytty\",\"database\":\"db_ziwfjgytty\"}]": "{\"transaction_id\":null,\"event_serial_no\":0,\"commit_lsn\":\"00000029:00000298:0003\",\"change_lsn\":\"NULL\"}"
    },
```

I'll revert this change for now